### PR TITLE
add new script to filter v2 variants

### PR DESCRIPTION
**Pull request by williamphu**
New script to filter v2 exome variants to those that have AC = 0 or NA in v2 genomes and v3 non_v2 genomes.

### CODE REVIEW

**Comment by williamphu** [comment]
1st run with Hail version 0.2.65-367cf1874d85

`hailctl dataproc submit wphu filter_v2_exomes_to_unique.py --overwrite`

File: filter_v2_exomes_to_unique.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    samples_with_variants = hl.read_table(args.sample_with_variants_path)
+    original_count = samples_with_variants.count()
+    v2_genomes = hl.read_table("gs://gcp-public-data--gnomad/release/2.1.1/ht/genomes/gnomad.genomes.r2.1.1.sites.ht")
+    v3_variants = hl.read_table("gs://gcp-public-data--gnomad/release/3.1.1/ht/genomes/gnomad.genomes.v3.1.1.sites.ht")
+    v2_liftover = hl.read_table("gs://gcp-public-data--gnomad/release/2.1.1/liftover_grch38/ht/exomes/gnomad.exomes.r2.1.1.sites.liftover_grch38.ht")
+    v2_liftover = v2_liftover.key_by("original_locus", "original_alleles")
+    v2_liftover_index = v2_liftover[samples_with_variants.locus, samples_with_variants.alleles]
+    samples_with_variants = samples_with_variants.annotate(liftover_locus = v2_liftover_index.locus, liftover_alleles = v2_liftover_index.alleles)
+    logger.info("Filtering variants not found in v2 genomes or v3_non_v2...")
+    idx = v2_genomes.freq_meta.collect()[0].index({'group': 'adj'})
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:24]:
```suggestion
    freq_adj_idx = v2_genomes_ht.freq_meta.collect()[0].index({'group': 'adj'})
```

File: filter_v2_exomes_to_unique.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    samples_with_variants = samples_with_variants.annotate(v2_genomes_adj_freq = v2_genomes[samples_with_variants.locus,samples_with_variants.alleles].freq[idx])
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:25]: Could this just be:
```suggestion
    ht = ht.annotate(v2_genomes_adj_freq = v2_genomes_ht[ht.key].freq[v2_genomes_adj_freq])
```

File: filter_v2_exomes_to_unique.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    v2_genomes_filtering_expr = (samples_with_variants.v2_genomes_adj_freq.AC<1) | ~hl.is_defined(samples_with_variants.v2_genomes_adj_freq.AC)
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:26]:
```suggestion
    v2_genomes_filtering_expr = (ht.v2_genomes_adj_freq.AC < 1) | hl.is_missing(ht.v2_genomes_adj_freq.AC)
```

File: filter_v2_exomes_to_unique.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    samples_with_variants = samples_with_variants.filter(v2_genomes_filtering_expr)
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:27]:
```suggestion
    ht = ht.filter(v2_genomes_filtering_expr)
```

File: filter_v2_exomes_to_unique.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    idx = v3_variants.freq_meta.collect()[0].index({'group': 'raw', 'subset': 'non_v2'})
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:28]: Do we want raw here, or adj? If adj:
```suggestion
    freq_adj_idx = v3_genomes_ht.freq_meta.collect()[0].index({'group': 'adj', 'subset': 'non_v2'})
```

File: filter_v2_exomes_to_unique.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    samples_with_variants = samples_with_variants.annotate(v3_non_v2_adj_freq = v3_variants[samples_with_variants.liftover_locus,samples_with_variants.liftover_alleles].freq[idx])
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:29]:
```suggestion
    ht = ht.annotate(v3_non_v2_adj_freq = v3_genomes_ht[ht.liftover_locus, ht.liftover_alleles].freq[freq_adj_idx])
```

File: filter_v2_exomes_to_unique.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    v3_non_v2_filtering_expr = (samples_with_variants.v3_non_v2_adj_freq.AC<1) | ~hl.is_defined(samples_with_variants.v3_non_v2_adj_freq.AC)
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:30]:
```suggestion
    v3_non_v2_filtering_expr = (ht.v3_non_v2_adj_freq.AC < 1) | hl.is_missing(ht.v3_non_v2_adj_freq.AC)
```

File: filter_v2_exomes_to_unique.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    samples_with_variants = samples_with_variants.filter(v3_non_v2_filtering_expr)
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:31]:
```suggestion
    ht = ht.filter(v3_non_v2_filtering_expr)
```

File: filter_v2_exomes_to_unique.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    logger.info(f"Writing Hail Table to {args.output_path_prefix}/v2_exomes_unique_samples_with_variants.ht")
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:32]:
```suggestion

    logger.info(f"Writing Hail Table to {args.output_path_prefix}/v2_exomes_unique_samples_with_variants.ht")
```

File: filter_v2_exomes_to_unique.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    samples_with_variants = samples_with_variants.checkpoint(f"{args.output_path_prefix}/v2_exomes_unique_samples_with_variants.ht", overwrite = args.overwrite)
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:33]:
```suggestion
    ht = ht.checkpoint(f"{args.output_path_prefix}/v2_exomes_unique_samples_with_variants.ht", overwrite = args.overwrite)
```

File: filter_v2_exomes_to_unique.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    logger.info(f"{original_count} variants were filtered to {samples_with_variants.count()} variants.")
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:34]:
```suggestion
    logger.info(f"{original_count} variants were filtered to {ht.count()} variants.")
```

File: filter_v2_exomes_to_unique.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    if args.overwrite:
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:35]:
```suggestion

    if args.overwrite:
```

File: filter_v2_exomes_to_unique.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        logger.info(f"Writing TSV to {args.output_path_prefix}/v2_exomes_unique_samples_with_variants.tsv")
+        samples_with_variants.export(
```
Comment (jkgoodrich) [inline filter_v2_exomes_to_unique.py:37]:
```suggestion
        ht.export(
```

**Review by jkgoodrich — COMMENTED** [review]
Just cosmetic changes, the actual filtering looks good to me

File: filter_v2_exomes_to_unique.py (hunk 5 above)
Comment (williamphu) [reply filter_v2_exomes_to_unique.py:28]: good catch!

**Comment by williamphu** [comment]
I had to port over the code here into the other branch so I could run both, so I'll close this PR, and new comments can go into the new PR.

**Pull request closed without merging**